### ISSUE TITLE
Sampling/shuffling sequences from a SequenceCollection or Alignment

### ISSUE THREAD

**Issue by dansondergaard** [biocore/scikit-bio#1122]
Hi,

This is just a request for an early code review. I have not yet been able to actually run the tests since I'm unable to install scikit-bio with my changes in a conda environment (complaints about freetype missing). Any suggestions as how to accomplish this will be greatly appreciated ;)

I have implemented sampling of sequences from a `SequenceCollection` or `Alignment`. This is especially nice to have when implementing bootstrapping or when an expensive analysis needs to be run on a smaller random subset of the `SequenceCollection`.

The implementation supports sampling without replacement (with replacement causes problems with duplicated sequence id's). By default, the sampling reduces to a simple shuffling of the sequences (rows) of the collection.

I hope that this feature is of interest to you :)

**Comment by jairideout** [comment]
Thanks for your interest in contributing this functionality to scikit-bio!

Re: install issues, I'm guessing the freetype complaint is happening when installing matplotlib. Have you tried installing matplotlib with conda? `conda install matplotlib`

Here's how I set up a development environment for scikit-bio (we'll update our contributing guide at some point to include conda-specific instructions):

1. create a new conda environment with Python 3.4: `conda create -n scikit-bio python=3.4`
2. activate the environment: `source activate scikit-bio`
3. cd into your clone of your scikit-bio fork
4. install all scikit-bio dependencies available on conda: `conda install --file ci/conda_requirements.txt`
5. install the rest of scikit-bio dependencies via pip: `pip install -r ci/pip_requirements.txt`
6. install scikit-bio itself in development mode (changes to code won't require a re-install each time): `pip install --no-deps -e .`
7. run the test suite: `make test`

Let me know how it goes! I'll follow up in a bit about the proposed addition of sampling to these classes.

**Comment by anderspitman** [comment]
:+1: I've been meaning to add this to our docs. See #1124

**Comment by jairideout** [comment]
@dansondergaard this sampling functionality sounds really useful! @ebolyen and I propose the following API for `TabularMSA`:

```python
def sample(self, n, replace=False) -> TabularMSA:
   ...
```

where `n` and `replace` match the parameter names used in `skbio.stats.subsample_counts`.

Since `TabularMSA` will replace `Alignment` in the near future we think this method makes sense there. Additionally, sampling with replacement is possible because a unique identifier (key) is optional on `TabularMSA`.

The future of `SequenceCollection` is unclear at this time so we want to avoid adding functionality to it. Potentially a function accepting an iterable of sequences and returning a list of sequences could be useful (this also doesn't suffer from the replacement problem). Regardless, `TabularMSA` has the clearest API at this time so we would like sampling implemented there first.

Does this API make sense to you? Is this something you are interested in working on?

**Comment by dansondergaard** [comment]
@jairideout Thank you for the instructions, they worked out perfectly! Much easier to test locally than with wait for CI ;)

I wasn't aware that `Alignment` was being replaced, but `TabularMSA` sounds like a worthy replacement. The API suggested makes good sense. My only concern is that I would also (at some point) like to be able to sample columns (positions). This could either be accomplished with an `axis` argument, as used in `numpy` or by having `sample_sequences` and `sample_positions` (or whatever columns are called in TabularMSA). That is why I named it `sample_sequences` in my PR.

I suggest to add the `axis` parameter to be consistent with the [TabularMSA RFC](https://github.com/biocore/scikit-bio-rfcs/blob/master/active/001-tabular-msa.md). I'll open a new PR as soon as I get to look at this (later this week, hopefully).

**Comment by ebolyen** [comment]
:+1: for `axis` parameter.

**Comment by jairideout** [comment]
:+1: for `axis`. Thanks for working on this @dansondergaard!

**Comment by dansondergaard** [comment]
@ebolyen @jairideout, great! Happy to contribute.

**Comment by dansondergaard** [comment]
Just realized that this depends on #1102 for slicing the `TabularMSA`. It is of course possible to implement sampling without slicing, but I think it's worth waiting a bit since the implementation will be a lot shorter and prettier.

What are your thoughts, @ebolyen, @jairideout?

**Comment by jairideout** [comment]
:+1: it's probably worth waiting for `__getitem__` (I think @ebolyen is working on this, or will start working on it soon)

**Comment by ebolyen** [comment]
Working on that for the next day or two, so that should be in soonish!

**Comment by anderspitman** [comment]
:+1:

**Comment by dansondergaard** [comment]
:+1:

**Comment by dansondergaard** [comment]
Closing this issue since it's no longer relevant to me and the PR isn't mergeable anymore.
